### ISSUE TITLE
Avoid creating channels when not having enough candidates.

### ISSUE THREAD

**Issue by hoplaaa** [ksedgwic/clboss#84]
Following investigation done in #81 

`ChannelCreator` builds channels lots to reduce onchain fees.
It calls the `Dowser` to do the channel sizing.
When there is not enough candidates, it allocates the leftover equally to the candidates:
https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L130-L155

As a consequence, we observed abnormally huge channels created to very small peers.
Instead, we could wait for having enough candidates for the available or create channels with their normal size and keep leftover onchain.

**Comment by ZmnSCPxj** [comment]
The reason for distributing the remaining amount is that it might end up in a situation where the remaining amount is too small for a decent channel (i.e. the remaining amount is less than the minimum channel size).  This will cause other mechanisms to attempt to move funds from offchain to onchain (`Boss::Msg::NeedsOnchainFunds`, `Boss::Mod::NeedsOnchainFundsSwapper`) so as to get enough onchain funds to make a decent channel size.

In particular, any solution here would interact with the minimum channel size.  The remaining amount may be just below the minimum channel size.  Indeed, one reason for stopping getting more candidates is because the remaining amount is no longer enough for a decent channel.

Here is a concrete example:

* We have a minimum channel size of 5mBTC.
* We have three candidates A B C, each has a dowsing result of 5.1mBTC.
* We have 15mBTC onchain.
* We take A and B and assign 5.1mBTC to them, so our remaining is 4.8mBTC.
* Our dilemma:
  * 4.8mBTC is below the minimum channel size, so assigning it to C is not economically useful.
  * Splitting it between A and B leads to channels which are almost +50% higher than their dowsed values, which is also not economically useful.

Currently the code takes the latter horn.  In this situation, is there a way out of this dilemma?

In any case, the OP complaint is that the actual amount is more than double the dowsed value.  Since at the above case the extremis is 50%, I think for a quick solution to this issue, we could check if distributing would cause one candidate to have more than 150% of its original dowsed amount.  That should still take the above latter horn in the dilemma without causing a livelock in the creator code.

**Comment by hoplaaa** [comment]
You are pointing a particular case, when `remaining < min_remaining`.
In that case, the current idea of splitting that small amount is good enough.

My problem is when `remaining > min_remaining`.
My concrete example being 100mBTC on chain, A(5.1mBTC) and B (5.1mBTC) the only available candidates, remaining 89.8mBTC shared between A and B.

> Since at the above case the extremis is 50%, I think for a quick solution to this issue, we could check if distributing would cause one candidate to have more than 150% of its original dowsed amount.

In that case we could have other problems, for example having 20Msat , A 8Msat,  B 8Msat, remaining 4 Msat. In such case I prefer to keep the 4Msat for new candidates.


Moving the following test _before trying to split the remaining_, to line 129 would be keeping fund onchain when having too much fund / not enough candidates.
https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L157-L159

Doing so, `ChannelCreator` would come back over and over as long as we have 2 new candidates in the list, creating too much onchain calls.

Instead we can wait for having enough candidates to minimalize the onchain fees by adding to the previous lines :
https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L119-L120

Doing shorter, we can add `|| remaining >= min_remaining` to the following test and remove L157-L159
https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L118-L122

This is possible because `remaining` beeing over `min_remaining` should normally be already spent on the last candidate :
https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L89-L90

**Comment by hoplaaa** [comment]
There is a problem though, already mentionned in #81.

At the early hours of the node life, when the node has no peers and no routing history, we need to allow `ChannelCreator` to quickly create few channels to first candidates, popular nodes.
In that case, we don't want to force `ChannelCreator` to wait for having enough (only most popular) candidate to use all the initial funding.

But, that is very specific case. Maybe we can deal with separately.
If not, we could come back to your option or to keeping `remaining` onchain and solving the issue of repetitive `ChannelCreator` calls.

**Comment by ZmnSCPxj** [comment]
> There is a problem though

Precisely why the code is done that way, with the `remaining >= min_remaining` check done *after* attempting to redistribute, it has to handle that early case. The idea is that after handling the early case, you will have lots of candidates already, so it is likely that you will be splitting any remainder among a larger set of candidates and having smaller amounts.

> Doing shorter, we can add `|| remaining >= min_remaining` to the following test and remove L157-L159

`result.clear()` will cause us to abort this attempt, meaning that we will leave the entire amount onchain and not open any channels, is that really what you are proposing?

**Comment by hoplaaa** [comment]
> The idea is that after handling the early case, you will have lots of candidates already, so it is likely that you will be splitting any remainder among a larger set of candidates and having smaller amounts.

Currently we cannot assume that we have enough candidates.
I got the experience on a ~1 month old node, fully using clboss since the beginning.
I understood that another user got the experience when launching clboss on an existing node. `ChannelCreator` might have been called too early and created 2 big channels to small nodes, using all available funds.

> `result.clear()` will cause us to abort this attempt, meaning that we will leave the entire amount onchain and not open any channels, is that really what you are proposing?

Yes.
We can also create the channels and keep the remaining onchain.
I'm fine with both. 
We could even go with some more hybrid approach : 
 - remaining very small -> split to candidates to avoid dust.
 - remaining relatively small -> do not create the channels and wait for more candidates.
 - remaining relatively big -> create the channel and keep the remaining on chain.

**Comment by ZmnSCPxj** [comment]
> We could even go with some more hybrid approach :

What are "very small", "relatively small" and "relatively big"?  How can we judge that while requiring as little extra information from the user as possible?  I think it would be a loss to *require* such settings from the user.  Default settings should "just work".  Make it changeable sure, but defaults should work for most people.

Note that we already set aside a reserve, that is what the `--clboss-min-onchain` setting is. The amount that the `ChannelCreator::Planner` object is given does not include this reserve.  Thus, dust outputs are not possible even with a tiny remaining amount, since it would be onsolidated with this `--clboss-min-onchain` reserve.

The main reason for avoiding making a small remaining amount is to prevent triggering the `NeedsOnchainFundsSwapper`.  The intent of that swapper is that if a channel of yours gets closed (e.g. by the peer, because we cannot control them) any liquidity from that channel should be put back into channels.  However, if the liquidity you got from the closed channel is small, it may not be practical to put it into a new, tiny channel.  So what `NeedsOnchainFundsSwapper` does is it triggers an offchain-to-onchain swap to get more onchain funds so that we have a practical amount onchain that can be put into channels.  That is the main reason why we have a `min_remaining` setting at all.

**Comment by ZmnSCPxj** [comment]
> remaining relatively small -> do not create the channels and wait for more candidates.

Do note that the initial planning loop has a `remaining < min_amount` check, too.  This means that if you have candidates A, B, C, D, E with dowsing results all equal to 5.1mBTC, and you have 15mBTC onchain, then `ChannelCreator` will never create channels --- once it assigns 5,1mBTC to A and B, what remains is 4.8mBTC, which is below the channel minimum amount, and the planner will now complete with only A=5.1mBTC and B=5.1mBTC and remaining=4.8mBTC.  If we do not create channels *now*, then the next time planning is invoked, then it will still have substantially the same inputs --- A, B, C, D, E, , G with A and B having dowsing results of 5.1mBTC, and remaining 4.8mBTC, which will again enter the "do not create" branch.

We need to ensure forward progress, and the splitting up logic here is part of that forward progress.

**Comment by ZmnSCPxj** [comment]
> Currently we cannot assume that we have enough candidates.
> I got the experience on a ~1 month old node, fully using clboss since the beginning.
I understood that another user got the experience when launching clboss on an existing node. ChannelCreator might have been called too early and created 2 big channels to small nodes, using all available funds.


If this "~1 month old node" is the same node that triggered #81, then merging in #82 will fix this --- `ChannelFinderByPopularity` in "aggressive" mode will look for 5 candidates, meaning you *will* get 5 candidates with fairly high probability of being large (since `ByPopularity` is a lottery weighted on node size).  So we *can* assume that.

The second case ("existing node") is not covered by that yet, but I think we can add a database flag where `ChannelFinderByPopularity` makes sure it *always* runs aggresively at least once, even if the node already has many channels.

**Comment by hoplaaa** [comment]
>What are "very small", "relatively small" and "relatively big"? How can we judge that while requiring as little extra information from the user as possible? I think it would be a loss to require such settings from the user. Default settings should "just work". Make it changeable sure, but defaults should work for most people.

I agree, we don't want to have options, it should work by itself.

We are at the step after the distribution of the available fund to the candidates is done.
Either there is no more funds or no more candidates : 
https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L72-L75

We want to know what to do with the remaining fund:
https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L123

The idea is to use what we have : the limit `min_amount`, the total funds available `total `and remaining  the `remaining`.

1. very small : `remaining `<  `min_amount `-> split to candidates to avoid dust
2. relatively small : `remaining `< 20% of `total `or < `min_remaining `-> do not create the channels and wait for more candidates
3. relatively big : `remaining `> 20% of `total `-> create the channel and keep the remaining on chain 

Edit : added or < `min_remaining `on option 2. to ensure we don't trigger `NeedsOnchainFundsSwapper`
Edit 2: for 2. and 3. 20-80% of `total `is arbitrary, it would work the opposite way or 50-50.
Edit 3: reason for 3 is to maintain the possibility to create first channels to popular peers at node creation

**Comment by hoplaaa** [comment]
> > Currently we cannot assume that we have enough candidates.
> > I got the experience on a ~1 month old node, fully using clboss since the beginning.
> > I understood that another user got the experience when launching clboss on an existing node. ChannelCreator might have been called too early and created 2 big channels to small nodes, using all available funds.
> 
> If this "~1 month old node" is the same node that triggered #81, then merging in #82 will fix this --- `ChannelFinderByPopularity` in "aggressive" mode will look for 5 candidates, meaning you _will_ get 5 candidates with fairly high probability of being large (since `ByPopularity` is a lottery weighted on node size). So we _can_ assume that.
> 
> The second case ("existing node") is not covered by that yet, but I think we can add a database flag where `ChannelFinderByPopularity` makes sure it _always_ runs aggresively at least once, even if the node already has many channels.

I'm sorry but I do not agree. 
`ChannelCreator` shouldn't assume that there is always have enough candidates. That can happen and that could even be good that this is acceptable. That would mean less pressure to have that list full, more time to find good candidates, possibility to empty or renew it, in general more freedom to evolve other modules. 

>In any case, the OP complaint is that the actual amount is more than double the dowsed value. Since at the above case the extremis is 50%, I think for a quick solution to this issue, we could check if distributing would cause one candidate to have more than 150% of its original dowsed amount. That should still take the above latter horn in the dilemma without causing a livelock in the creator code.

I understand and I'm good with your proposal too.
Making my own proposals helped me to dig the code. By the way, thanks for all explanations.

**Comment by ZmnSCPxj** [comment]
Let us stew this topic a little more.  I am concerned about livelock, and I need to think a bit more on *why* I put the `remaining > min_remaining` check *after* the try-to-split.  My initial reaction to (re-)reading this code was "that is strange", so there must have been a reason for it, but unfortunately that idiot ZmnSCPxj from a year ago did not think to put a comment there, so I need to dig through old memory files to see why I did that, or if it was just because I was not paying attention.

**Comment by hoplaaa** [comment]
>I am concerned about livelock

You are right, we might loop waiting for more candidates. 

>I need to think a bit more on why I put the remaining > min_remaining check after the try-to-split.

I think you said it earlier, the idea was to keep low onchain costs.
The split fails when having huge remaining, when "it's worth" another onchain transaction.

**Comment by ZmnSCPxj** [comment]
> The split fails when having huge remaining, when "it's worth" another onchain transaction.

But that is what the `remaining > min_remaining` check is for --- if you leave more than `min_remaining` that "it is worth" the separate onchain tx to just channel now and wait for more candidates later.... I think.  Stupid ZmnSCPxj one year ago, not leaving comments in code...

**Comment by hoplaaa** [comment]
What if `remaining > min_remaining ` was not added intentionally _after the split_, but intentionally added _before the deduction_ that comes after.
We don't need to get rid of `remaining` when `remaining > min_remaining` :

https://github.com/ZmnSCPxj/clboss/blob/00cdded89cd5eac01ee5c6bb4675ecc7fa20778f/Boss/Mod/ChannelCreator/Planner.cpp#L157-L177

**Comment by ZmnSCPxj** [comment]
Maybe?  I just have this nagging vague memory that long ago, the `remaining >= min_remaining` check was *before* the split logic, then I thought of something, then I moved it to *after* the split logic.  Unfortunately the records end there, there is no comment as to why I did that, and the memory is vague and might therefore be from bitrot rather than an accurate memory of what happened.

It seems to me that the simplest solution is to move it to before the splitting logic.
